source/fuzz/fuzzer.cpp

```diff
@@ -165,200 +268,46 @@ bool Fuzzer::ApplyPassAndCheckValidity(FuzzerPass* pass) const {
-    // There is a theoretical possibility that no pass instances were created
-    // until now; loop again if so.
-  } while (pass_instances.GetPasses().empty());
-
-  RepeatedPassRecommenderStandard pass_recommender(&pass_instances,
-                                                   fuzzer_context_.get());
-
-  std::unique_ptr<RepeatedPassManager> repeated_pass_manager = nullptr;
-  switch (repeated_pass_strategy_) {
-    case RepeatedPassStrategy::kSimple:
-      repeated_pass_manager = MakeUnique<RepeatedPassManagerSimple>(
-          fuzzer_context_.get(), &pass_instances);
-      break;
-    case RepeatedPassStrategy::kLoopedWithRecommendations:
-      repeated_pass_manager =
-          MakeUnique<RepeatedPassManagerLoopedWithRecommendations>(
-              fuzzer_context_.get(), &pass_instances, &pass_recommender);
-      break;
-    case RepeatedPassStrategy::kRandomWithRecommendations:
-      repeated_pass_manager =
-          MakeUnique<RepeatedPassManagerRandomWithRecommendations>(
-              fuzzer_context_.get(), &pass_instances, &pass_recommender);
-      break;
+  if (num_repeated_passes_applied_ >=
+      fuzzer_context_->GetTransformationLimit()) {
+    return {FuzzerResultStatus::kFuzzerStuck, {}};
   }
 
   do {
     if (!ApplyPassAndCheckValidity(
-            repeated_pass_manager->ChoosePass(transformation_sequence_out_))) {
-      return {Fuzzer::FuzzerResultStatus::kFuzzerPassLedToInvalidModule,
-              std::vector<uint32_t>(), protobufs::TransformationSequence()};
+            repeated_pass_manager_->ChoosePass(transformation_sequence_out_))) {
+      return {Fuzzer::FuzzerResultStatus::kFuzzerPassLedToInvalidModule, {}};
     }
-  } while (ShouldContinueFuzzing());
+  } while (ShouldContinueFuzzing(initial_num_of_transformations,
+                                 num_of_transformations_to_apply));
 
-  // Now apply some passes that it does not make sense to apply repeatedly,
-  // as they do not unlock other passes.
-  std::vector<std::unique_ptr<FuzzerPass>> final_passes;
-  MaybeAddFinalPass<FuzzerPassAdjustBranchWeights>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassAdjustFunctionControls>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassAdjustLoopControls>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassAdjustMemoryOperandsMasks>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassAdjustSelectionControls>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassAddNoContractionDecorations>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassInterchangeSignednessOfIntegerOperands>(
-      &final_passes);
-  MaybeAddFinalPass<FuzzerPassInterchangeZeroLikeConstants>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassPermutePhiOperands>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassSwapCommutableOperands>(&final_passes);
-  MaybeAddFinalPass<FuzzerPassToggleAccessChainInstruction>(&final_passes);
-  for (auto& pass : final_passes) {
+  for (auto& pass : final_passes_) {
```

Comment:
A reader might wonder why we don't try to respect "num_of_transformations_to_apply_" here.

I see a few choices:

1. Document this to acknowledge that we deliberately don't respect it.
2. Start the loop with a check that if this value is non-zero we haven't exceeded it, breaking the loop if so
3. Guard the loop body with a call to ShouldContinueFuzzing

I think I favour option 1: these are special transformations that it doesn't make sense to do repeatedly, but we don't want to radically diminish our chances of doing them at all when we are tint-fuzzing, and I fear that options 2 and 3 would lead to that.

Reply:
Fixed in 1f68fec.